finat/finiteelementbase.py

```diff
@@ -144,6 +145,112 @@ def point_evaluation(self, order, refcoords, entity=None):
         :param entity: the cell entity on which to tabulate.
         '''
 
+    @abstractproperty
+    def dual_basis(self):
+        '''Returns a tuple where each element of the tuple represents one
+        functional in the dual space. Each functional is represented by
+        a tuple of tuples containing the points (PointSet), a weight tensor
+        which holds the weights for each component of value_shape of functions,
+        and a alpha tensor for extracting the alpha components from ReferenceGrad,
+        sorted by total derivative order.
+
+        For example, a dual basis containing 2 functionals with maximum derivative
+        order of 1 would be represented by:
+
+        .. highlight:: python
+        .. code-block:: python
+
+            ((((point_set_10, weight_tensor_10, alpha_tensor_10),
+               (point_set_11, weight_tensor_11, alpha_tensor_11)), None)
+             (((point_set_20, weight_tensor_20, alpha_tensor_20),
+              ()), None)
```

Comment:
This is not really a datastructure.

Reply:
I agree that it's quite ugly. Do you have anything in mind which could be iterated over and be easy to create other dual bases from?

Reply:
So I think I don't understand what the bits are and how you destructure them. Each dual basis is made of, I think, a "quadrature rule" (point set + weight expression pair) and a multiindex indicating which derivatives that evaluation needs. Then there's something with components?